src/sysbench.py

```diff
@@ -0,0 +1,282 @@
+            and not self.is_running()
+            and not self.is_failed()
+        )
+
+    def is_failed(self) -> bool:
+        """Checks if the sysbench service has failed."""
+        return self.is_prepared() and os.path.exists(self.svc_path) and service_failed(self.svc)
+
+    def run(self) -> bool:
+        """Run the sysbench service."""
+        if self.is_stopped() or self.is_failed():
+            return service_restart(self.svc)
+        return self.is_running()
+
+    def stop(self) -> bool:
+        """Stop the sysbench service."""
+        if self.is_running():
+            return service_stop(self.svc)
+        return self.is_stopped()
+
+    def unset(self) -> bool:
+        """Unset the sysbench service."""
+        try:
+            result = self.stop()
+            result ^= service_stop(self.ready_target)
+            os.remove(f"/etc/systemd/system/{self.ready_target}")
+            os.remove(self.svc_path)
+            return daemon_reload() and result
+        except Exception:
+            pass
+
+
+class SysbenchStatus:
+    """Renders the sysbench status updates the relation databag."""
+
+    def __init__(self, charm: ops.charm.CharmBase, relation: str, svc: SysbenchService):
+        self.charm = charm
+        self.svc = svc
+        self.relation = relation
+
+    @property
+    def _relation(self) -> Dict[str, Any]:
+        return self.charm.model.get_relation(self.relation)
+
+    def app_status(self) -> SysbenchExecStatusEnum:
+        """Returns the app status."""
+        if not self._relation:
+            return None
+        return SysbenchExecStatusEnum(
+            self._relation.data[self.charm.app].get("status", SysbenchExecStatusEnum.UNSET.value)
+        )
+
+    def unit_status(self) -> SysbenchExecStatusEnum:
+        """Returns the unit status."""
+        if not self._relation:
+            return None
+        return SysbenchExecStatusEnum(
+            self._relation.data[self.charm.unit].get("status", SysbenchExecStatusEnum.UNSET.value)
+        )
+
+    def set(self, status: SysbenchExecStatusEnum) -> None:
+        """Sets the status in the relation."""
+        if not self._relation:
+            return
+        if self.charm.unit.is_leader():
+            self._relation.data[self.charm.app]["status"] = status.value
+        self._relation.data[self.charm.unit]["status"] = status.value
+
+    def _has_error_happened(self) -> bool:
+        for unit in self._relation.units:
+            if self._relation.data[unit].get("status") == SysbenchExecStatusEnum.ERROR.value:
+                return True
+        return (
+            self.unit_status() == SysbenchExecStatusEnum.ERROR
+            or self.app_status() == SysbenchExecStatusEnum.ERROR
+        )
+
+    def service_status(self) -> SysbenchExecStatusEnum:
+        """Returns the status of the sysbench service."""
+        if not self.svc.is_prepared():
+            return SysbenchExecStatusEnum.UNSET
```

Comment:
Curious here: Couldn't there be a case where the service fails before being prepared? Having  `self.svc.is_failed()` true and `self.svc.is_prepared()`false. IIUC it would get status `UNSET` instead of `ERROR`. Is this expected?

Reply:
oh ok, makes sense. Thx!